pkg/extra/handlers.go

```diff
@@ -63,13 +92,67 @@ func (a *API) handleConsent(w http.ResponseWriter, r *http.Request) {
 
 }
 
-func NewAPI(service ServiceInterface, kratos kratos.ServiceInterface, logger logging.LoggerInterface) *API {
+func (a *API) shouldEnforceWebAuthnWithSession(ctx context.Context, session *client.Session) (bool, error) {
+	hasOIDC := false
+	// enforce only if one of the authentication methods was oidc
+	for _, method := range session.AuthenticationMethods {
+		if method.Method != nil && *method.Method == "oidc" {
+			hasOIDC = true
+		}
+	}
+
+	if !hasOIDC {
+		return false, nil
+	}
+
+	webAuthnAvailable, err := a.kratos.HasWebAuthnAvailable(ctx, session.Identity.GetId())
+	if err != nil {
+		return false, err
+	}
+
+	return !webAuthnAvailable, nil
+}
+
+func (a *API) webAuthnSettingsRedirect(w http.ResponseWriter, returnTo string) {
+	redirect, err := url.JoinPath("/", a.contextPath, "/ui/setup_passkey")
+	if err != nil {
+		err = fmt.Errorf("unable to build webauthn redirect path, possible misconfiguration, err: %v", err)
+		a.logger.Error(err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	errorId := "session_aal2_required"
+
+	// Set the original consent URL as return_to, to continue the flow after WebAuthn key is set
+	redirectTo, _ := httpHelpers.AddParamsToURL(redirect, httpHelpers.QueryParam{Name: "return_to", Value: returnTo})
+
+	w.WriteHeader(http.StatusSeeOther)
+	_ = json.NewEncoder(w).Encode(
+		kratos.ErrorBrowserLocationChangeRequired{
+			Error:             &client.GenericError{Id: &errorId},
+			RedirectBrowserTo: &redirectTo,
+		},
+	)
```

Comment:
IMO we should either respond with a 30x and set the `Location` in the header OR respond with a 40x (maybe 422 as kratos does?) and set the redirect location in the response body. Considering this is a backchannel request I would go with the latter.

Reply:
Keep in mind this is not a backchannel request since it comes from a frontend useragent and it's returned to it
This is also aligned with how we've done redirect in both admin ui and login ui

Reply:
Ok maybe I abused the term back-channel. But it is an XHR request, meaning that it does not make sense to return a redirect. Our approach right now is not consistent, in some places we return `200`, in others `303` and in others `403`.

Out of the 3 I think that `303` is the one that feels most un-intuitive. But we can move it to a new issue to implement a uniform strategy for that.

Reply:
Definitely
IMO 303 would still be the most "appropriate", but I see what you mean. Probably 200 or 422, but let's move it to an issue. We'll have to address the smaller things at some point